common/network-common/src/main/java/org/apache/spark/network/server/TransportRequestHandler.java

```diff
@@ -181,6 +182,17 @@ public void onFailure(Throwable e) {
   private void processStreamUpload(final UploadStream req) {
     assert (req.body() == null);
     try {
+      // Retain the original metadata buffer, since it will be used during the invocation of
+      // this method. Will be released later.
+      req.meta.retain();
+      // Make a copy of the original metadata buffer. In benchmark, we noticed that
+      // we cannot respond the original metadata buffer back to the client, otherwise
+      // in cases where multiple concurrent shuffles are present, a wrong metadata might
+      // be sent back to client. This is related to the eager release of the metadata buffer,
+      // i.e., we always release the original buffer by the time the invocation of this
+      // method ends, instead of by the time we respond it to the client. This is necessary,
+      // otherwise we start seeing memory issues very quickly in benchmarks.
+      ByteBuffer meta = cloneBuffer(req.meta.nioByteBuffer());
```

Comment:
I am sorry but I fail to see how concurrent shuffle is generating such an error.  
As `req` is final method argument I am sure this inner method cannot see another instance, so is it possible the problem is just a an index change within `req.meta` which can be solved by some duplicate methods? 

Can we have a unit test for this method to investigate this a bit more?

Reply:
I have thought about this a lot and I do not think we need to send back the meta as I think it could be made easily available at the client side.  

So the complete use case starts with sending of `UploadStream` by calling `TransportClient#uploadStream`. Here we are already store something (the callback) for each request ID which identify uniquely the request:

https://github.com/apache/spark/blob/7b48c50b88f43a548bb845ca994b9e8794ef593e/common/network-common/src/main/java/org/apache/spark/network/client/TransportClient.java#L223

From the meta what we really use is the index but even if you need the whole `PushBlockStream` as metadata you can store it these for the request ID. This way when RpcFailure comes back with the request ID we can simply use the stored data.

I suggest to extend the `TransportClient#uploadStream` arguments with the necessary data (index) and store this data for the requestID NOT the meta as a `ManagedBuffer`. 
This new parameter could be an object and null by default. 

As `uploadStream` is used only two places in production this is not a huge change. 

This way you can get rid not only the metadata send back but the `cloneBuffer` and ` JavaUtils.encodeHeaderIntoErrorString` and `BlockPushException#decodeException` methods too.

Reply:
For the `req.meta` issue, my understanding is the following:
`processStreamUpload` is only responsible for creating a a `StreamCallbackWithID` to be added into the FrameDecoder as a stream interceptor.
The Netty ByteBuf `req.meta` will be released by the time this method exits.
However, the stream callback would need to respond `req.meta` after this method exits.
Accessing the value of the Netty ByteBuf after it's released is what's causing the issue mentioned in the comment.
I tried to delay the release of `req.meta` until the stream callback finishes processing the stream, however that can lead to memory issues on the executor side when there are many blocks to be transferred.

Reply:
I still do not want to change the `TransportClient#uploadStream` API itself.
This transport layer utility was previously used for transferring large RDD partition blocks, and now reused for doing block push.
In the future, it is possible that other use cases might benefit from this utility as well.
I believe keeping this API generic and not specific to one use case is important.

For the change you proposed to keep the `PushBlockStream` as metadata tracked on the client side, I also thought about doing that during implementation.
It's cleaner than the current approach.

One way to do this without incurring any potential protocol change would be to make `BlockPushCallback` inside `OneForOneBlockPusher` stateful.
Currently, that callback is stateless, so multiple invocations to `TransportClient#uploadStream` for the same batch of blocks would reuse the same callback object.
If we make that callback object stateful, to keep track of the additional metadata ManagedBuffer, then the callback object would have what we need built into it during object creation.
My concern during the implementation was the potential JVM pressure this approach might generate, since we will create one callback object per block to be pushed.

What do you think?
Also CC @mridulm @tgravescs @squito @Ngone51 @jiangxb1987 for your inputs on this.

Reply:
> I tried to delay the release of req.meta until the stream callback finishes processing the stream, however that can lead to memory issues on the shuffle service side when there are many blocks to be transferred.

I still do not see how a cloned buffer which should live also until the stream callback finishes processing the stream would help on this. The `ByteBuf` is reference counted so with a retain & release the same time period could be covered without the duplication. The duplication even increases the memory footprint for the time when both instances are alive.

Reply:
Netty `ByteBuf` is reference counted, and will be deallocated once `release` calls decreases the reference to 0.
Nio `ByteBuffer` is not, and it's Java reference based and will be deallocated by GC.
My original attempt was to use retain & release to cover a longer time period so the `ByteBuf` would live until the callback finishes processing the stream.
However, during stress test where we stress a single shuffle service with many concurrent block push requests, we found out that the shuffle service can fail to allocate additional Netty `ByteBuf` if we delay the release of `req.meta` buffer.
I can update this PR with that version of the code which does not clone the `ByteBuf` into a `ByteBuffer` and also attach the error log we saw, in case there is something missed on my end.

Reply:
Let's wait the decision about the idea to store the indexes at client side as with that the whole problem disappears.

Reply:
Makes sense.
One minor correction, we would need both the index and the blockId from `PushBlockStream` on the client side.
The index is needed for handling push failure while the blockId is needed for handling push success (inside `BlockPushCallback`).

CC @mridulm @tgravescs @squito @Ngone51 @jiangxb1987 again for inputs on this discussion.

Reply:
@attilapiros 
Why do you think that caching the metadata on the client is a better approach than sending this metadata back with the response?

Reply:
@otterc 
https://github.com/apache/spark/pull/29855#discussion_r498985394

> It's cleaner than the current approach.

https://github.com/apache/spark/pull/29855#discussion_r498714105

> This way you can get rid not only the metadata send back but the cloneBuffer and JavaUtils.encodeHeaderIntoErrorString and BlockPushException#decodeException methods too.

For me `JavaUtils.encodeHeaderIntoErrorString` and `BlockPushException#decodeException` is quite painful as they are bit too hacky: sending back the meta data within a String along with the the exceptions stack. I understand the reason behind. This way such a generic and basic class as `RpcFailure` could be kept simple and still reused. But we have a cleaner solution at hand.

Reply:
@Victsm I was looking this a bit more and I have a new idea.

Let's create new `BlockPushCallback` instance for each block in `OneForOneBlockPusher#start`. Extend the `BlockPushCallback` constructor with the block ID and with the index and access it from there. 

This way `TransportClient#uploadStream` can remain as it is. What do you think?

Reply:
@attilapiros this is the approach I mentioned above, along with the concern about no longer able to reuse callback object.
Let me try code it that way and benchmark it to see if this concern is real or not.

> One way to do this without incurring any potential protocol change would be to make BlockPushCallback inside OneForOneBlockPusher stateful.
Currently, that callback is stateless, so multiple invocations to TransportClient#uploadStream for the same batch of blocks would reuse the same callback object.
If we make that callback object stateful, to keep track of the additional metadata ManagedBuffer, then the callback object would have what we need built into it during object creation.
My concern during the implementation was the potential JVM pressure this approach might generate, since we will create one callback object per block to be pushed.

Reply:
I think the major issue here is that, we still need the metadata on the client side the handle push callbacks. It sounds like a good idea to keep the metadata inside the `BlockPushCallback`, so we don't need to worry about the lifecycle of the metadata.

Reply:
Another possible solution is to reuse the fetch logic instead of push, but that requires a much wider change, and also changes the control flow, thus I'm hesitate to propose the change.

Reply:
@jiangxb1987 what do you think about the need of creating one callback object per block?
On the fetch side, `ChunkCallback` is reused, and only `DownloadCallback` is created one per block.
However, `DownloadCallback` is only used with large blocks, so it won't be created for every block.

Reply:
Are we really adding a lot of memory pressure by not reusing the callback? I'm expecting a size of ~10KB for each BlockPushCallback, so it should be fine if we have hundreds or thousands of them. If the size is well over hundreds of KBs then I'd assume something went wrong.

Reply:
Do we have cases where large number of callback's are created which is causing memory pressure/gc issues ?

Reply:
Spark does not do this previously, i.e. creating large number of callbacks.
This occurs in the Spark executor side, not in the shuffle service side though.
We are in the process of benchmarking this to see if there's any visible impact.

Reply:
An update on this thread.
We have made the suggested change and benchmarked it.
The benchmark stress tested a single shuffle service by pushing a total of 15M 10KB blocks from 200 concurrent clients, thus each client would need to push 75K blocks (with 75K callback objects created).
Here're the results:
On the server side, after removing the need to clone buffer, we saw a small dip in CPU utilization (~6%) and physical memory utilization (~10%)
On the client side, the additional callback objects created per block do not have a visible impact on the CPU and memory utilizations, most likely due to these callback objects being short-lived.

With these results, we are confident with taking this suggested approach.